### PR TITLE
EL-2457 client date of birth editing

### PR DESCRIPTION
**Pull request by sryalls**
[MCC: Change client date of birth](https://dsdmoj.atlassian.net/browse/EL-2457)

## What changed and Why?

This PR implements the ability to edit a client's date of birth in the case details workflow. The change includes:
- Express Schema Validation: Validation logic now uses express schema validation for the date of birth fields. This ensures type safety, declarative validation rules, and consistent error handling. The schema covers required fields, valid date formats, impossible dates, and unchanged data, reducing duplication and improving maintainability.
- New controller logic for rendering the edit form, validating user input, and updating client details via API.
Validation is performed in the controller and helpers, covering required fields, valid date formats, impossible dates, and unchanged data.
- Error handling uses GOV.UK error summary and inline messages, preserving form data and CSRF tokens for user experience and security.
- The Nunjucks template leverages GOV.UK date input and error summary components for accessibility and consistency.
Integration tests cover all workflows: successful edits, validation errors, impossible dates, unchanged data, API errors, and error propagation.
- No separate middleware is used for DoB validation; all logic is encapsulated in the controller and helpers.
This implementation ensuring robust validation, error handling, and workflow integration.

### Custom error vs JSON error
This PR upgrades validation for the Date of Birth form from a simple, ad-hoc JSON error approach (used in phone number validation) to a robust, type-safe schema using GovUkValidationError. The new schema supports error prioritization, AC-compliant messaging, and advanced UI logic.

Why?
The previous JSON error method was quick and easy for simple forms but lacked type safety, error prioritization, and extensibility.
The new Date of Birth schema is more complex but provides consistent, maintainable, and extensible error handling, making it ideal for forms with complex business rules and advanced UI requirements.
This change ensures validation is reliable, future-proof, and aligned with best practices for complex workflows.

I believe the complexity on error prioritisation and field highlighting would have been more challenging with the JSON approach and future use cases, Address in particular will benefit from the additional framework complexity of the custom error based solution. 

## Guidance to review

Check that the edit form renders correctly and displays errors as expected.
Confirm that validation covers all edge cases (missing fields, invalid dates, unchanged data).
Ensure API integration updates client details and handles errors gracefully.
Review integration tests for coverage of all workflows and error scenarios.
Validate that documentation and code changes are consistent with the technical implementation guide and consolidation plan.

### Validation Pattern
This implementation uses express-validator middleware with priority-based error filtering, which differs from the email validation pattern that uses a [validateForm()](vscode-file://vscode-app/private/var/folders/w2/n07yw7cx7_97x_8ljh5j_wmh0000gp/T/AppTranslocation/E5909845-210C-4EEA-BAC7-95032FE37946/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) interface.

The date validation requires complex priority-based filtering (e.g., show "missing day" but not "invalid format" errors simultaneously), which is well-suited to the middleware approach. The existing implementation provides:

Priority-based error filtering (missing fields > format errors > business logic errors)
Multiple same-priority errors (e.g., multiple missing fields)
Comprehensive validation (format, real date, future date, unchanged detection)
Question for Review: Should we standardize on the email validation pattern ([validateForm()](vscode-file://vscode-app/private/var/folders/w2/n07yw7cx7_97x_8ljh5j_wmh0000gp/T/AppTranslocation/E5909845-210C-4EEA-BAC7-95032FE37946/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) interface) across all form validations, or maintain different patterns for different complexity levels? The current date validation would require significant refactoring to fit the email pattern.

### Downstream server mocking 
[editWorkflow.spec.ts](https://github.com/ministryofjustice/laa-manage-your-civil-cases/pull/183/files#diff-ae5eaf2fc5ac88fb19cf32e13dfb1adc7ae25b8dd604f6b3a9ec686e14c13fa8) implements a mock of the interface layer to the downsrteam serve. This may be relevant to know future considerations. 

### i18n Implementation Recommendation
It would be advisable to implement internationalization to support language compliance, including Welsh, and storing copy text in the repo in a more organized way. During this feature development I tried integrating validation messages with potential multi-language support and ran into difficulties with ES2020 and ES2022 compatibility issues so decided resolving these was out of scope for this feature.

### Recommend to do the following in the near future:

- Add i18next middleware to Express.js app with locale detection from URL parameters 
- Create centralized translation files (locales/en-GB/ and locales/cy-GB/) to replace hardcoded strings 
- Update validation helpers to use translation keys instead of static error messages 
- Resolve ES module compatibility between i18next and existing TypeScript configuration 
- Implement locale-aware routing with Welsh Language Standards compliance

 ### Moment.js
Currently using Moment.js for date handling. While functional and reliable, Moment.js is in maintenance mode. Consider migrating to date-fns as future proofing. 

## Checklist

Before you ask people to review this PR:

- [x] **Tests and linting** are passing.  
- [x] **Branch is up to date** with `main` (no merge conflicts).  
- [x] **No unnecessary whitespace changes** (avoid unnecessary diffs).  
- [x] **PR description clearly explains** what changed and why, with a JIRA ticket/Trello link.  
- [x] **Diff has been checked** for any unexpected changes.  
- [x] **Commit messages are clear** and explain what will change, if individual commit, needs to be revisited retroactively.

### CODE REVIEW

File: src/helpers/dateValidationHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import type { DateFields } from '#src/types/dateTypes.js';
+
+export type { ValidationError, DateFields };
+
+/**
+ * Safely convert unknown value to string
+ * @param {unknown} value - The value to convert to string
+ * @returns {string} String representation of the value
+ */
+export function safeToString(value: unknown): string {
```
Comment (imtiazAhmd) [inline src/helpers/dateValidationHelpers.ts:19]: Can we use existing `safeString` helper from `dataTransformers` ?

File: routes/editClientDetails.ts [hunk 2]
```diff
@@ -8,6 +9,8 @@ import {
   postEditClientPhoneNumber
 } from '#src/scripts/controllers/editClientDetailsController.js';
 import { validateEditClientPhoneNumber } from '#src/middlewares/phoneNumberSchema.js';
+import { getEditDateOfBirth, postEditDateOfBirth } from '#src/scripts/controllers/editDateOfBirthController.js';
+import { dateOfBirthSchema } from '#src/validation/index.js';
```
[the remark this answers is not in the document]
Comment (sryalls) [reply routes/editClientDetails.ts:13]: I think that would probably be the best approach

File: views/case_details/edit-date-of-birth.njk [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        <!-- CSRF Protection -->
+        <input type="hidden" name="_csrf" value="{{ csrfToken }}" />
+        
+        <!-- Original values for change detection -->
+        <input type="hidden" name="originalDay" value="{{ originalData.day }}" />
+        <input type="hidden" name="originalMonth" value="{{ originalData.month }}" />
+        <input type="hidden" name="originalYear" value="{{ originalData.year }}" />
+        
+        {{ govukDateInput({
+          id: "dateOfBirth",
```
[the remark this answers is not in the document]
Comment (sryalls) [reply views/case_details/edit-date-of-birth.njk:44]: With reference to the full stop:  ‘For example, 27 3 2007.’ hint text - include the full stop <- from the AC

File: src/helpers/dateValidationHelpers.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+ * 
+ * When multiple validation errors exist, this function filters them to show
+ * only the errors with the highest priority (lowest numeric value) to avoid
+ * overwhelming the user with too many error messages.
+ * 
+ * @param {ValidationError[]} errors Array of validation errors from express-validator
+ * @param {Record<string, number>} errorPriorityMap Map of error messages to priority numbers (lower = higher priority)
+ * @returns {ValidationError[]} Filtered array containing only highest priority errors
+ */
+export function filterErrorsByPriority(
```
[the remark this answers is not in the document]
Comment (sryalls) [reply src/helpers/dateValidationHelpers.ts:97]: With the date errors, there are some errors where multiple might apply but there's only one that's directly relevant. It's not directly in the AC but kind of comes out as implied when there are multiple errors with different sub-fields (i.e we have month 13 and a blank day). Since only one of the errors can be put in the inline error message, we decide which is the most directly actionable and show that.

File: src/validation/schemas/dateOfBirthSchema.ts [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+/**
+ * @file Date of Birth Schema Validation - Simplified Inline Logic
+ * 
+ * This schema provides direct custom validation functions for each date field,
+ * eliminating the need for complex abstraction layers. All validation logic
+ * is consolidated directly in the schema for improved maintainability.
+ * 
+ * **Key Features:**
+ * - Direct inline validation logic (no external validator dependencies)
```
[the remark this answers is not in the document]
Comment (sryalls) [reply src/validation/schemas/dateOfBirthSchema.ts:25]: Yeah. that's probably worth looking at.

File: src/helpers/dateValidationHelpers.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (sryalls) [reply src/helpers/dateValidationHelpers.ts:97]: @MazOneTwoOne - After taking another look yesterday evening, the prioritisation can be simplified. The structure is based on the, incorrect, assumption that more of the fields in the application would have similar constraints. Since it's only date fields it's applicable to, this way of doing it is overkill and, yeah, I can simplify it.

File: views/case_details/edit-date-of-birth.njk [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+              value: formData.year,
+              attributes: {
+                "aria-describedby": "dateOfBirth-error" if highlightYear,
+                "autocomplete": "off"
+              }
+            }
+          ]
+        }) }}
+
+        {{ govukButton({
```
[the remark this answers is not in the document]
Comment (sryalls) [reply views/case_details/edit-date-of-birth.njk]: Fixed.
